Formula/postgresql@12.rb

```diff
@@ -81,22 +81,100 @@ def post_install
     return if ENV["CI"]
 
     (var/"log").mkpath
-    (var/"postgres").mkpath
-    unless File.exist? "#{var}/postgres/PG_VERSION"
-      system "#{bin}/initdb", "--locale=C", "-E", "UTF-8", "#{var}/postgres"
+    versioned_postgresql_data_dir.mkpath
+    unless versioned_data_dir_exists?
+      system "#{bin}/initdb", "--locale=C", "-E", "UTF-8", versioned_postgresql_data_dir
     end
   end
 
+  # Previous versions of this formula used the same data dir as the regular
+  # postgresql formula. So we check whether the versioned data dir exists
+  # and has a PG_VERSION file, which should indicate that the versioned
+  # data dir is in use. Otherwise, returns the old data dir path.
+  def postgresql_datadir
+    if versioned_data_dir_exists?
+      versioned_postgresql_data_dir
+    else
+      old_postgresql_data_dir
+    end
+  end
+
+  def versioned_postgresql_data_dir
+    "#{var}/#{name}"
+  end
+
+  def old_postgresql_data_dir
+    "#{var}/postgres"
+  end
+
+  # Same as with the data dir - use old log file if the old data dir
+  # is version 12
+  def postgresql_log_path
+    if versioned_data_dir_exists?
+      "#{var}/log/#{name}"
+    else
+      "#{var}/log/postgres"
+    end
+  end
+
+  def versioned_data_dir_exists?
+    File.exist?("#{versioned_postgresql_data_dir}/PG_VERSION")
+  end
+
+  def conflicts_with_postgresql_formula?
+    Formula["postgresql"].any_version_installed?
+  end
+
+  # Figure out what version of PostgreSQL the old data dir is
+  # using
+  def old_postgresql_datadir_version_12?
+    File.exist?("#{old_postgresql_data_dir}/PG_VERSION") &&
+      File.read("#{old_postgresql_data_dir}/PG_VERSION").chomp == "12"
+  end
+
   def caveats
-    <<~EOS
-      To migrate existing data from a previous major version of PostgreSQL run:
-        brew postgresql-upgrade-database
+    caveats = ""
+
+    # Check if we need to print a warning re: data dir
+    if old_postgresql_datadir_version_12?
+      caveats += if conflicts_with_postgresql_formula?
+        # Both PostgreSQL and PostgreSQL@12 are installed
+        <<~EOS
+          Previous versions of this formula used the same data directory as
+          the regular PostgreSQL formula. This causes a conflict if you
+          try to use both at the same time.
+
+          In order to avoid this conflict, you should make sure that the
+          #{name} data directory is located at:
+
+            #{versioned_postgresql_data_dir}
+
+        EOS
+      else
+        # Only PostgreSQL@12 is installed, not PostgreSQL
+        <<~EOS
+          Previous versions of #{name} used the same data directory as
+          the postgresql formula. This will cause a conflict if you
+          try to use both at the same time.
 
+          You can migrate to a versioned data directory by running this command:
+
+            mv -v "#{old_postgresql_data_dir}" "#{versioned_postgresql_data_dir}"
+
+          (Make sure PostgreSQL is stopped before executing this command)
+
```

Comment:
```suggestion
          (Make sure PostgreSQL is stopped before executing this command)
```

Reply:
The reason I have the extra line in here is to create extra spacing with the later caveats string. Without it, it would "stick" together with the `This formula has created…` stuff, which is always printed. And I figured doing it this way would be more sensible than creating a two-element array, and joining that, or something.

Reply:
Gotcha, fine with me 👍🏻